packages/lit-html/src/hydrate.ts

```diff
@@ -0,0 +1,407 @@
+ * nested templates, as repeat() creates. We need the stack to contain all
+ * NodeParts for that.
+ */
+type NodePartState =
+  | {
+      type: 'leaf';
+      /** The NodePart that the result is rendered to */
+      part: NodePart;
+    }
+  | {
+      type: 'iterable';
+      /** The NodePart that the result is rendered to */
+      part: NodePart;
+      value: Iterable<unknown>;
+      iterator: Iterator<unknown>;
+      done: boolean;
+    }
+  | {
+      type: 'template-instance';
+      /** The NodePart that the result is rendered to */
+      part: NodePart;
+
+      result: TemplateResult;
+
+      /** The TemplateInstance created from the TemplateResult */
+      instance: TemplateInstance;
+
+      /**
+       * The index of the next Template part to be hydrated. This is mutable and
+       * updated as the tree walk discovers new part markers at the right level in
+       * the template instance tree.  Note there is only one Template part per
+       * attribute with (one or more) bindings.
+       */
+      templatePartIndex: number;
+
+      /**
+       * The index of the next TemplateInstance part to be hydrated. This is used
+       * to retrieve the value from the TemplateResult and initialize the
+       * TemplateInstance parts' values for dirty-checking on first render.
+       */
+      instancePartIndex: number;
+    };
+
+/**
+ * hydrate() operates on a container with server-side rendered content and
+ * restores the client side data structures needed for lit-html updates
+ * such as TemplateInstances and Parts.
+ *
+ * hydrate() must be called on DOM that adheres the to lit-ssr structure for
+ * parts. NodeParts must be represented with both a start and end comment
+ * marker, and NodeParts that contain a TemplateInstance must have the template
+ * digest written into the comment data.
+ *
+ * Since render() encloses its output in a NodePart, there must always be a root
+ * NodePart.
+ *
+ * Example (using for # ... for annotations in HTML)
+ *
+ * Given this input:
+ *
+ *   html`<div class=${x}>${y}</div>`
+ *
+ * The SSR DOM is:
+ *
+ *   <!--lit-part AEmR7W+R0Ak=-->  # Start marker for the root Node Part created
+ *                                 # by render(). Includes the digest of the
+ *                                 # template
+ *   <div class="TEST_X">
+ *     <!--lit-bindings 0--> # Indicates there are attribute bindings here
+ *                           # The number is the depth-first index of the parent
+ *                           # node in the template.
+ *     <!--lit-part-->  # Start marker for the ${x} expression
+ *     TEST_Y
+ *     <!--/lit-part-->  # End marker for the ${x} expression
+ *   </div>
+ *
+ *   <!--/lit-part-->  # End marker for the root NodePart
+ *
+ * @param rootValue
+ * @param container
+ * @param userOptions
+ */
+export const hydrate = (
+  rootValue: unknown,
+  container: Element | DocumentFragment,
+  options: Partial<RenderOptions> = {}
```

Comment:
Oh yeah, it's a little tricky because it's non-trivial to know where to start the hydrate treewalk from, isn't it.

Opened https://github.com/Polymer/lit-html/issues/1383.